### PR TITLE
Subclassing Validator

### PR DESCRIPTION
**Pull request by amymok**
The scope of work is described here: https://github.com/18F/data-federation-project/issues/167

## Additions:
- Created a `ValidatorOutput` class to better organize, structure, and standardize the output
  - This include standardizing the error definition, what should be included
- Added abstract methods `validate` in `Validator` and `evaluate` in `RowwiseValidator` to force implementation for all children classes
- Added tests to test `ValidatorOutput`, which include functions that were moved into the class and not tested before
- Added docstrings to most methods (those method that have been added or changed)

## Changes:
- Replaced output creations that exist in different places with the use of `ValidatorOutput`
- Made `combine_validation_results` as a method in `ValidationOutput` called `combine`
- Broke out `row_validation_errors`
  - Since it is only being called at one place right now, moved it back to where it is called, and made use of the `ValidationOutput.add_row_error` to take care of the formatting/structuring
  - The message interpolation broke out to its own function within `RowwiseValidator`
- Moved `rows_from_source` into `Validator` 
- Rearranged the `ingestors.py` file, so helper functions exist together, and classes group together
- Moved `rows_from_source` into `Validator`

## Removals:
- Removed `count_valid_rows` as `ValidatorOutput` will now handle it
  - This function added an extra set of validation counts to the output, which makes the output confusing and causes errors on the counts

## Tests:
- Ran all examples manually to make sure the UI didn't break because of these changes
- Ran a few tests on FNS validation data service to make sure changes didn't break the application there
- Ran the automated tests

## TODOs:
- Will finish these in a separate PR
  - Break methods/functions into smaller pieces
  - Clarification on settings (Need some more time to understand them)
  - Managing Validators
  - More documentation (in terms of markdown to help developers understand how to use these)

### CODE REVIEW

File: data_ingest/ingestors.py [hunk 1]
```diff
@@ -378,69 +493,166 @@ def cast_value(value):
...
-        rows = []
-        (table['headers'], numbered_rows) = rows_from_source(source)
         for (rn, row) in numbered_rows.items():
 
             # This is to remove the header row
             if rn == UPLOAD_SETTINGS['OLD_HEADER_ROW']:
                 continue
 
-            errors = []
+            # errors = []
```
Comment (tbaxter-18f) [inline data_ingest/ingestors.py]: Did you want to leave this commented out?

File: examples/p02_budgets/README.md [hunk 2]
```diff
@@ -1,17 +1,17 @@
 
 # Budgets example app
 
-Demonstrates `data_ingest` with some basic [customizations](../../docs/custom.md)
+Demonstrates `data_ingest` with some basic [customizations](../../docs/customize.md)
```
Comment (tbaxter-18f) [inline examples/p02_budgets/README.md:4]: were these links wrong before, or is there a new page I'm not seeing?

Comment (amymok) [reply examples/p02_budgets/README.md:4]: It was linked incorrectly before.

File: data_ingest/ingestors.py (hunk 1 above)
Comment (amymok) [reply data_ingest/ingestors.py]: Oops, will remove

**Review by tbaxter-18f — APPROVED** [review]
LGTM!

**Pull request merged**